compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -391,6 +400,44 @@ fn link_rlib<'a, B: ArchiveBuilder<'a>>(
     ab
 }
 
+/// Extract all symbols defined in raw-dylib libraries, collated by library name.
+///
+/// If we have multiple extern blocks that specify symbols defined in the same raw-dylib library,
+/// then the CodegenResults value contains one NativeLib instance for each block.  However, the
+/// linker appears to expect only a single import library for each library used, so we need to
+/// collate the symbols together by library name before generating the import libraries.
+fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(Symbol, Vec<DllImport>)> {
+    let mut dylib_table: FxHashMap<Symbol, FxHashSet<Symbol>> = FxHashMap::default();
+
+    for lib in used_libraries {
+        if lib.kind == NativeLibKind::RawDylib {
+            let name = lib.name.unwrap_or_else(||
+                bug!("`link` attribute with kind = \"raw-dylib\" and no name should have caused error earlier")
+            );
+            let imports = dylib_table.entry(name).or_insert_with(FxHashSet::default);
+            for imported_symbol in &lib.dll_imports {
+                imports.insert(imported_symbol.name);
+            }
+        }
+    }
+
+    // FIXME: when we add support for ordinals, fix this to propagate ordinals.  Also figure out
+    // what we should do if we have two DllImport values with the same name but different
+    // ordinals.
+    dylib_table
+        .drain()
```

Comment:
```suggestion
        .into_iter()
```

Reply:
I've made the change, but could you explain why `into_iter` is preferable here?

Reply:
`drain` is a more "exotic" and special-purposed method, it leaves `dylib_table` usable but empty (which we don't need here).
`into_iter`/`iter`/`iter_mut` are regular methods for everyday use, used by `for` loops etc.

Reply:
Cool, thanks!  I hadn't picked up on that difference between those methods.